### ISSUE TITLE
Working now

### ISSUE THREAD

**Issue by GeorgeS2019** [kaiidams/Voice100Sharp#1]
I tried before the recent code update. 
Now working fine, both tts and stt (ASR)

Do you have suggestion how to improve the speech recognition accuracy?

Do you know DeepSpeech (PyTorch)? I wonder if it is possible to convert the model to ONNX.

**Comment by kaiidams** [comment]
ASR needs more improvements. If we wants to detect commands then we can try keyword detection like [this](https://www.tensorflow.org/tutorials/audio/simple_audio). [YamNet ](https://www.tensorflow.org/hub/tutorials/yamnet)is for TensorFlow lite, but I have [its rewrite for Unity](https://github.com/kaiidams/YamNetUnityDemo).

If we wants ASR really, we can try NeMo. NVIDIA has small ASR models. https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/models.html#quartznet .

**Comment by GeorgeS2019** [comment]
@kaiidams  can your TTS solution with ONNX create [a file](https://github.com/David-Ochoa/godot-lipsync/blob/master/Speach/merchant_female.json) like this

**Comment by kaiidams** [comment]
Voice100 and QuartzNet both produce raw English texts but not pronunciations. If we need lip syncing,
- the model needs to predict pronunciations
- and the model must predicts from small time frames so that the time and sound are aligned.

I'll study what we can do.

This is ONNX conversion of NeMo QuartzNet.

https://gist.github.com/kaiidams/baa13dcec194516c0bbb850de811cda5

The transcript below is from Bloomberg Podcast today. I think it is reasonably good.

> Recognized as: that does not seem to have come pass this year indeed divon the main object of chines bulling is polian democratic beinat and onby the sintoken america's rhetorical promotion octomol received andsnot resonate wiing ete asian licoso ranrk

**Comment by kaiidams** [comment]
If we need quick solution, then we can use the output of align model

```
        public byte[] Speak(string text)
        {
            long[] encoded = _encoder.Encode(text);
            if (encoded.Length == 0) return new byte[0];
            long[] aligned = Align(encoded);
            var y = Predict(aligned);
            return MemoryMarshal.Cast<short, byte>(y).ToArray();
        }
```

aligned is a sequence of characters like this. One character is 10ms.

```
_____tha_t___  doe_s___  n_ot__       _s_e_em____    to     ha_ve    _c_o_me_  _____p_a_s_s______  this__  _year_________        __in___d_e_ed________________  ____d_i______v_o_n_____________ __________the___ _m_ain____  _ob___j_e_ct   of   _chi_n____e_ss_ b_u_l_l_inng _iss  ____p_o___l_i___an__    de___m_o__cra____t_i_c__ _b_e_________________i___n_a___t___________________  _and__ __o_n__b_y___ the_  s_iin______t_o_____k_e_n__________      _a___m_e_r__i__c_a_'''_s__   __rhe___t_o__r__i__c_al__  pro__m_o___tion_  _oc___t_o___m_o_l__ _re____c_eiv_e_d_________  ___a_n___d___s__n_o_t______      ____r_e___s_o___n_a_tee__ _w_i_______ing____ ___e____t_e____________  ___a____sia_n__ l_i____c_o______s_o_________ _rra_n________r_k__________
```

**Comment by kaiidams** [comment]
`alignedText` is a sequence of characters, each character represents what is spoken in `audio` for each 10ms time step.

https://github.com/kaiidams/Voice100Godot/commit/0071709ef07467903f72342983b50cfdb1903fe7

**Comment by GeorgeS2019** [comment]
Output, 10ms.

_/_/_/_/_/HH/HH/_/EH1/EH1/_/_/_/_/L/L/_/OW0/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/AE1/_/_/M/M/AA1/_/_/_/_/_/_/R/AA1/_/_/_/_/_/_/_/_/K/_/AH0/_/_/_/_/_/_/T/_/_/_

**Comment by GeorgeS2019** [comment]
@kaiidams 

I wonder if you have plan to improve the accuracy of ASR? The vad has noticeably improved the responsiveness. 

Here is the Microsoft [documentation ](https://docs.microsoft.com/en-us/azure/cognitive-services/speech-service/how-to-custom-speech-train-model)on use cases where accuracy can be improved.

Typically for Windows Speech Recognition, we read a few texts a few times to improve the recognition. In your implementation, will that involve a pipeline of using python script in Voice100 for e.g.[ Voice Conversion](https://paperswithcode.com/paper/voice-conversion-can-improve-asr-in-very-low), followed by exporting the model for .NET consumption. 

Even for your current implementation, it is possible to do regex to retrieve the right commands, to address a few characters that are miss-spelled. 

Great job!

**Comment by kaiidams** [comment]
@GeorgeS2019 Currently the ASR model was trained with clean-100 of https://www.openslr.org/12/. I'll try clean-360 in the beginning. But it may not be as good as QuartzNet as the model size is much smaller.

**Comment by kaiidams** [comment]
@GeorgeS2019 Now the new ASR model was trained with `clean-360`, the model seems better but not as good as QuartzNet. You can see the predicted results in `README.md`.

**Comment by GeorgeS2019** [comment]
@kaiidams 

Ideally, the speaker read the test text and record into wav files. Then the training compared the wav profiles created by the speaker as well as those used for testing, to identify the key differences and make corrections to minimize mistakes. 

Just a suggestion.

After that. I think there are tricks which big Tech companies used to improve accuracy.
e.g. Natural Language Generation method to replace words which make no sense with statistically more common expression.

**Comment by GeorgeS2019** [comment]
@kaiidams 

I have followed that you have been [busy ](https://github.com/kaiidams/voice100) , any chance you will be interested in updating this project with your latest work?
